### PR TITLE
Use ids to validate translations

### PR DESCRIPTION
**Pull request by mkangia**
https://dimagi-dev.atlassian.net/browse/ICDS-679
This is a complicated change and i have tried to make this easy to understand.

Current brief:
When a user uploads app translations, we let them compare that with the current app translations and share diff.
Currently this is done by using sheet_names i.e `Menus_and_forms`, `menuN`, `menuN_formM` to compare content of the excel sheet uploaded and the excel sheet that would be generated for the current app.
In case the module and form structure is different this would lead to incorrect module/form being compared to. Earlier it was not expected to happen but it is happening now.

This PR updates the logic to 
1. find the id of module/form from the uploaded "Menus_and_forms" sheet
2. use this module/form id to find the translations for the module/form in the current app
3. the headers for the current app translations are also mapped to id now

Terms `current` and `expected` signify the current app translations and can/will be merged to just `current`.

FF: `VALIDATE_APP_TRANSLATIONS`

fyi @esoergel

### CODE REVIEW

File: corehq/apps/translations/validator.py [hunk 1]
```diff
@@ -31,8 +31,8 @@ class UploadedTranslationsValidator(object):
     def __init__(self, app, uploaded_workbook, lang_to_compare, lang_prefix='default_'):
         self.app = app
         self.uploaded_workbook = uploaded_workbook
-        self.headers = None
-        self.expected_rows = None
+        self.expected_headers = dict()  # module_or_form_id: list of headers
+        self.expected_rows = dict()  # module_or_form_id: translations
```
Comment (mkangia) [inline corehq/apps/translations/validator.py]: Comments are what it should be changed to.

File: corehq/apps/translations/validator.py [hunk 2]
```diff
@@ -57,46 +64,77 @@ def _generate_expected_headers_and_rows(self):
...
+            self.current_sheet_name_to_module_or_form_type_and_id[row[sheet_name_column_index]] = Unique_ID(
+                row[type_column_index],
+                row[unique_id_column_index]
+            )
 
     @memoized
-    def _get_header_index(self, sheet_name, header):
-        for index, _column_name in enumerate(self.headers[sheet_name]):
+    def _get_current_header_index(self, module_or_form_id, header):
+        for index, _column_name in enumerate(self.expected_headers[sheet_name]):
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/translations/validator.py]: Updated later in [commit](https://github.com/dimagi/commcare-hq/pull/24837/commits/72e1a5aec106e7f8a7f847743f18f20a28111fc4) to correct var

File: corehq/apps/translations/validator.py [hunk 3]
```diff
@@ -56,9 +62,38 @@ def _generate_expected_headers_and_rows(self):
...
+        type_column_index = self._get_current_header_index(MODULES_AND_FORMS_SHEET_NAME, 'Type')
+        for row in all_module_and_form_details:
+            self.current_sheet_name_to_module_or_form_type_and_id[row[sheet_name_column_index]] = Unique_ID(
+                row[type_column_index],
+                row[unique_id_column_index]
+            )
 
     @memoized
-    def _get_header_index(self, sheet_name, header):
+    def _get_current_header_index(self, module_or_form_id, header):
```
Comment (mkangia) [inline corehq/apps/translations/validator.py]: uses of the method updated in a later commit, bad miss.

**Review by kaapstorm — APPROVED** [review]
Nice. Nothing blocking.

File: corehq/apps/translations/validator.py [hunk 4]
```diff
@@ -32,8 +36,9 @@ class UploadedTranslationsValidator(object):
     def __init__(self, app, uploaded_workbook, lang_to_compare, lang_prefix='default_'):
         self.app = app
         self.uploaded_workbook = uploaded_workbook
-        self.headers = None
-        self.expected_rows = None
+        self.uploaded_headers = dict()  # sheet_name: list of headers
+        self.current_headers = dict()  # module_or_form_id: list of headers
+        self.current_rows = dict()  # module_or_form_id: translations
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/translations/validator.py]: i think i am fine with this naming here, its respecting the naming being followed for all object properties. I picked the other names from the changes you made earlier :)
I will make a follow up PR to combine these usages, if it's possible.
Explicit naming is great and i am a big fan of it but i try to keep a tab on their length as well :)

File: corehq/apps/translations/validator.py [hunk 5]
```diff
@@ -45,87 +50,138 @@ def __init__(self, app, uploaded_workbook, lang_to_compare, lang_prefix='default
         self.app_translation_generator = AppTranslationsGenerator(
             self.app.domain, self.app.get_id, None, self.app.default_language, target_lang,
             self.lang_prefix)
+        self.current_sheet_name_to_module_or_form_type_and_id = dict()
+        self.uploaded_sheet_name_to_module_or_form_type_and_id = dict()
 
-    def _generate_expected_headers_and_rows(self):
-        self.headers = {h[0]: h[1] for h in get_bulk_app_sheet_headers(
+    def _generate_current_headers_and_rows(self):
+        self.current_headers = {h[0]: h[1] for h in get_bulk_app_sheet_headers(
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/translations/validator.py]: sounds good, updated

**Comment by mkangia** [comment]
Deployed on staging for testing by GS(Sankalp/Akshita)

**Comment by kaapstorm** [comment]
:tada: Nice to get this merged!

**Pull request merged**